tests/unit/test_reservation.py

```diff
@@ -208,5 +215,402 @@ def test___heterogenous_session_infos___create_sessions___grouped_by_instrument_
+        assert [conn.site for conn in connections] == [2, 5]
+        assert [conn.channel_name for conn in connections] == ["3", "6"]
+        assert [conn.session for conn in connections] == [nifoo_info.session, nibar_info.session]
+
+
+def test___wrong_session_type___get_connection___type_error_raised(
+    session_management_client: Mock,
+) -> None:
+    with ExitStack() as stack:
+        grpc_session_infos = create_nifake_session_infos(1)
+        grpc_session_infos[0].channel_mappings.add(pin_or_relay_name="Pin1", site=2, channel="3")
+        reservation = MultiSessionReservation(session_management_client, grpc_session_infos)
+        session_info = stack.enter_context(reservation.create_session(_construct_session, "nifake"))
+
+        with pytest.raises(TypeError) as exc_info:
+            _ = reservation.get_connection(int)
+
+        assert "Incorrect type for session 'MySession0'." in exc_info.value.args[0]
+        assert (
+            "Expected <class 'int'>, got <class 'tests.utilities.fake_driver.Session'>."
+            in exc_info.value.args[0]
+        )
+
+
+def test___session_not_created___get_connection___type_error_raised(
+    session_management_client: Mock,
+) -> None:
+    grpc_session_infos = create_nifake_session_infos(1)
+    grpc_session_infos[0].channel_mappings.add(pin_or_relay_name="Pin1", site=2, channel="3")
+    reservation = MultiSessionReservation(session_management_client, grpc_session_infos)
+
+    with pytest.raises(TypeError) as exc_info:
+        _ = reservation.get_connection(fake_driver.Session)
+
+    assert "Incorrect type for session 'MySession0'." in exc_info.value.args[0]
+    assert (
+        "Expected <class 'tests.utilities.fake_driver.Session'>, got <class 'NoneType'>."
+        in exc_info.value.args[0]
+    )
+
+
+def test___session_not_created___get_connection_with_session_type_object___connection_returned(
+    session_management_client: Mock,
+) -> None:
+    grpc_session_infos = create_nifake_session_infos(1)
+    grpc_session_infos[0].channel_mappings.add(pin_or_relay_name="Pin1", site=2, channel="3")
+    reservation = MultiSessionReservation(session_management_client, grpc_session_infos)
+
+    connection = reservation.get_connection(object)
+
+    assert connection.pin_or_relay_name == "Pin1"
+    assert connection.site == 2
+    assert connection.channel_name == "3"
+    assert connection.session is None
+
+
+def test___reservation_order___get_connections_with_specified_order___connections_returned_in_specified_order(
+    session_management_client: Mock,
+) -> None:
+    grpc_session_infos = _create_grpc_session_infos_for_ordering()
+    reservation = MultiSessionReservation(
+        session_management_client,
+        grpc_session_infos,
+        reserved_pin_or_relay_names=["Pin3", "Pin1", "Pin4", "Pin2"],
+        reserved_sites=[1, 0],
+        reserved_instrument_type_ids=["nifoo", "nibar"],
+    )
+
+    nifoo_connections = reservation.get_connections(
```

Comment:
@pbirkhol-ni It's currently not an error to get connections for instrument_type_id=nifoo with pins that only match instrument_type_id=nibar, as long as some of the pins match nifoo. Please let me know if this is inconsistent with LV.

It now occurs to me that this also tolerates entirely made-up pin names, as long as some of the pins are real and match the session info. Perhaps we should validate that the pins and sites exist in the reservation data.

Reply:
Yes, this is inconsistent. In LabVIEW, if you get connections for a pin that does not have any connections, an error is thrown.

Reply:
I updated the code to error if the caller specifies pins, sites, or instrument type id and they don't match anything.